### PR TITLE
Add enable_cache parameter to APIRequestHelper

### PR DESCRIPTION
**Pull request by nvonpentz**
Passing the `enable_cache` parameter will not add the BYPASS_CACHE and DISABLE_CACHE load flags to the request which were previously set for all requests.

I tested with a local server that the browser caches and uses cached responses so long as the Cache-Control headers are set correctly and when these two flags are removed.

This PR does not change any actual usage of the api_request_helper, just adds the ability to cache. So nothing material should change after this.

Resolves https://github.com/brave/brave-browser/issues/29071

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Review by simonhong — APPROVED** [review]

File: components/api_request_helper/api_request_helper_unittest.cc [hunk 1]
```diff
@@ -54,13 +55,21 @@ class ApiRequestHelperUnitTest : public testing::Test {
...
           url_loader_factory_.ClearResponses();
           EXPECT_EQ(request.url, expected_url);
           EXPECT_EQ(request.method, expected_method);
+          if (enable_cache) {
+            EXPECT_EQ(request.load_flags, net::LOAD_DO_NOT_SAVE_COOKIES);
+          } else {
+            EXPECT_EQ(request.load_flags, net::LOAD_DO_NOT_SAVE_COOKIES |
+                                              net::LOAD_BYPASS_CACHE |
+                                              net::LOAD_DISABLE_CACHE);
+          }
```
Comment (yrliou) [inline components/api_request_helper/api_request_helper_unittest.cc:72]: Can we have some test case that actually sets enable_cache to true?
As far as I can see, we only executes the enable_cache = false path in all tests here.

**Review by cypt4 — CHANGES_REQUESTED** [review]

File: components/api_request_helper/api_request_helper.h [hunk 2]
```diff
@@ -106,6 +107,7 @@ class APIRequestHelper {
       const std::string& payload,
       const std::string& payload_content_type,
       bool auto_retry_on_network_change,
+      bool enable_cache,
```
[the remark this answers is not in the document]
Comment (yrliou) [reply components/api_request_helper/api_request_helper.h]: We should have a comment with an issue opened in the header to clearly saying this.

File: components/api_request_helper/api_request_helper.h [hunk 3]
```diff
@@ -100,6 +101,9 @@ class APIRequestHelper {
   // into string before validating the response. For these purposes
   // conversion_callback is added which receives raw response and can perform
   // necessary conversions.
+  //
+  // This method will be deprecated soon,
```
Comment (yrliou) [inline components/api_request_helper/api_request_helper.h:105]: // This method will be deprecated soon, please use the one underneath with APIRequestOption parameter.
  // https://github.com/brave/brave-browser/issues/29611

File: components/api_request_helper/api_request_helper.h [hunk 4]
```diff
@@ -132,7 +132,6 @@ class APIRequestHelper {
                   const std::string& payload,
                   const std::string& payload_content_type,
                   bool auto_retry_on_network_change,
-                  bool enable_cache,
```
Comment (yrliou) [inline components/api_request_helper/api_request_helper.h]: What's the reason that we shouldn't add enable_cache option here?

**Review by cypt4 — APPROVED** [review]
LGTM

**Pull request merged**